### PR TITLE
make miri happy

### LINKED ISSUE #29: miri stacked borrows error
When running the tests with `cargo +nightly miri test` it errors out with 

```
Undefined Behavior: attempting a read access using <441222> at alloc162558[0x10], but that tag does not exist in the borrow stack for this location
```

I think this is 'mostly' harmless as stacked borrows are overly restrictive. There are some ways to convince miri that it is correct, these may make the code a bit uglier, I may look into that soon. Eventually this should be fixed to be able to run the testsuite under miri anyways.

```
   --> src/lib.rs:522:13
    |
522 |             std::ptr::copy(src_ptr, dst_ptr, orig_len - seg_offset - 1);
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |             |
    |             attempting a read access using <441222> at alloc162558[0x10], but that tag does not exist in the borrow stack for this location
    |             this error occurs as part of an access at alloc162558[0x10..0x20]
    |
    = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
    = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
help: <441222> was created by a SharedReadWrite retag at offsets [0x0..0x10]
   --> src/lib.rs:520:27
    |
520 |             let dst_ptr = &mut self.segments[seg_idx][seg_offset] as *mut T;
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: BACKTRACE (of the first span):
    = note: inside `SegVec::<tests::DropCount<'_, i32>>::remove` at src/lib.rs:522:13: 522:72
note: inside `<Drain<'_, tests::DropCount<'_, i32>, mem_config::Exponential<1>> as std::iter::Iterator>::next`
   --> src/lib.rs:949:24
    |
949 |             let next = self.inner.remove(self.index);
    |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: inside `<&mut Drain<'_, tests::DropCount<'_, i32>, mem_config::Exponential<1>> as std::iter::Iterator>::next` at /home/ct/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:4010:9: 4010:24
    = note: inside `<&mut Drain<'_, tests::DropCount<'_, i32>, mem_config::Exponential<1>> as std::iter::Iterator>::fold::<(), [closure@std::iter::Iterator::for_each::call<tests::DropCount<'_, i32>, fn(tests::DropCount<'_, i32>) {std::mem::drop::<tests::DropCount<'_, i32>>}>::{closure#0}]>` at /home/ct/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:2480:29: 2480:40
    = note: inside `<&mut Drain<'_, tests::DropCount<'_, i32>, mem_config::Exponential<1>> as std::iter::Iterator>::for_each::<fn(tests::DropCount<'_, i32>) {std::mem::drop::<tests::DropCount<'_, i32>>}>` at /home/ct/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:856:9: 856:31
note: inside `<Drain<'_, tests::DropCount<'_, i32>, mem_config::Exponential<1>> as std::ops::Drop>::drop`
   --> src/lib.rs:981:9
    |
981 |         self.for_each(drop);
    |         ^^^^^^^^^^^^^^^^^^^
    = note: inside `std::ptr::drop_in_place::<Drain<'_, tests::DropCount<'_, i32>, mem_config::Exponential<1>>> - shim(Some(Drain<'_, tests::DropCount<'_, i32>, mem_config::Exponential<1>>))` at /home/ct/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:497:1: 497:56
    = note: inside `std::mem::drop::<Drain<'_, tests::DropCount<'_, i32>, mem_config::Exponential<1>>>` at /home/ct/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/mem/mod.rs:987:24: 987:25
note: inside `tests::test_drain`
   --> src/tests.rs:274:5
    |
274 |     drop(d);
    |     ^^^^^^^
note: inside closure
   --> src/tests.rs:253:17
    |
252 | #[test]
    | ------- in this procedural macro expansion
253 | fn test_drain() {
    |                 ^
    = note: this error originates in the attribute macro `test` (in Nightly builds, run with -Z macro-backtrace for more info)

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to previous error
```

### PR DESCRIPTION
**Pull request by vxpm**
fixes #29 and others

it would be good to add  `// SAFETY: ...` comments - i'll probably add some in a few hours.

### CODE REVIEW

File: src/tests.rs [hunk 1]
```diff
@@ -341,13 +353,19 @@ fn test_drain() {
...
+
             let d = sv.drain((Bound::Included(1), Bound::Included(i)));
             let drained = d.map(|i| i.1).collect::<Vec<_>>();
             assert_eq!(dc.get(), i, "i={}", i);
             assert_eq!(sv.len(), 8 - i, "i={}", i);
             assert_eq!(drained, (2..=(i as i32 + 1)).collect::<Vec<_>>(), "i={}", i);
         }
     }
+
+    for dc in to_cleanup.into_iter() {
```
Comment (mccolljr) [inline src/tests.rs:366]: This is fine, but I don't think it is strictly necessary - the memory allocated for these leaked `Box` values is negligible, and this code only runs within the tests, and this cleanup code won't run if the test fails anyway.

File: src/tests.rs [hunk 2]
```diff
@@ -613,8 +631,10 @@ fn test_segmented_iter() {
 
 #[test]
 fn test_sort() {
+    const COUNT: usize = if cfg!(miri) { 32 } else { 1000 };
```
Comment (mccolljr) [inline src/tests.rs:634]: No issues here, just wondering if this is a "takes too long" issue or a "miri actually breaks here" issue (I haven't run it in a long time)

**Review by mccolljr — APPROVED** [review]

**Comment by mccolljr** [comment]
`// SAFETY` comments would be nice, but I also think that your changes leave it fairly obvious why each new `unsafe` usage is safe.

Do you have any interest in modifying the CI code to run the tests under `miri` in addition to the current runs? Totally fine if not, I can try to get this enabled later this week, but since you're already here I thought I would ask.

File: src/tests.rs (hunk 1 above)
Comment (vxpm) [reply src/tests.rs:366]: i agree - it's just that miri will complain about leaked values if these are not deallocated

File: src/tests.rs (hunk 2 above)
Comment (vxpm) [reply src/tests.rs:634]: takes too long, yeah. the 32 is arbitrary, however

**Comment by vxpm** [comment]
sure! i'll do it together with the safety comments in a few hours.

**Pull request merged**